tree/treeplayer/inc/ROOT/TDataFrame.hxx

```diff
@@ -709,169 +717,131 @@ private:
       unsigned int nSlots = df->GetNSlots();
-      return SimpleAction<BranchType, ActionResultType, ActionType, decltype(this)>::BuildAndBook(this, bl, r, nSlots);
+      return BuildAndBook<BranchType>(bl, r, nSlots, (ActionType*)nullptr);
    }
 
-   // Do type guessing
-   template <ROOT::Internal::EActionType ActionType, typename ActionResultType>
+   // User did not specify type, do type guessing
+   template <typename ActionType, typename ActionResultType>
    TActionResultProxy<ActionResultType>
-   CreateAction(const BranchNames & bl, const std::shared_ptr<ActionResultType>& r,
+   CreateAction(const BranchNames& bl, const std::shared_ptr<ActionResultType>& r,
                 ROOT::Detail::TDataFrameGuessedType* /*dummy*/)
    {
       // More types can be added at will at the cost of some compilation time and size of binaries.
-      using ART_t = ActionResultType;
-      using TT_t = decltype(this);
-      const auto at = ActionType;
+      using AT_t = ActionType;
       auto df = GetDataFrameChecked();
       unsigned int nSlots = df->GetNSlots();
 
-
       auto tree = static_cast<TTree*>(df->GetDirectory()->Get(df->GetTreeName().c_str()));
       auto theBranchName = bl[0];
       auto branch = tree->GetBranch(theBranchName.c_str());
 
       if (!branch) {
          // temporary branch
          const auto &type_id = df->GetBookedBranch(theBranchName).GetTypeId();
-         if (type_id == typeid(char)) {
-            return SimpleAction<char, ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots);
-         } else if (type_id == typeid(int)) {
-            return SimpleAction<int, ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots);
-         } else if (type_id == typeid(double)) {
-            return SimpleAction<double, ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots);
-         } else if (type_id == typeid(std::vector<double>)) {
-            return SimpleAction<std::vector<double>, ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots);
-         } else if (type_id == typeid(std::vector<float>)) {
-            return SimpleAction<std::vector<float>, ART_t, at, TT_t>::BuildAndBook(this, bl, r, nSlots);
-         }
+         if (type_id == typeid(char)) { return BuildAndBook<char>(bl, r, nSlots, (AT_t*)nullptr); }
+         else if (type_id == typeid(int)) { return BuildAndBook<int>(bl, r, nSlots, (AT_t*)nullptr); }
+         else if (type_id == typeid(double)) { return BuildAndBook<double>(bl, r, nSlots, (AT_t*)nullptr); }
+         else if (type_id == typeid(std::vector<double>)) { return BuildAndBook<std::vector<double>>(bl, r, nSlots, (AT_t*)nullptr); }
+         else if (type_id == typeid(std::vector<float>)) { return BuildAndBook<std::vector<float>>(bl, r, nSlots, (AT_t*)nullptr); }
       }
       // real branch
       auto branchEl = dynamic_cast<TBranchElement *>(branch);
       if (!branchEl) { // This is a fundamental type
-         auto title    = branch->GetTitle();
+         auto title = branch->GetTitle();
```

Comment:
I propose to refine this aspect of the type guessing logic in another PR as the current one is not related to this particular aspect.